feat: Add command line/environment flag for commonConfig

ejlee3 opened the pull request:
<!-- Expected Commit Message Description (imported automatically by GitHub) -->
<!-- Must conform to [conventional commits guidelines](https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/.github/Contributing.md) -->
<!-- Expected Commit message must contain Closes/Fixes #IssueNumber statement when there is a related issue -->

<!-- Add additional detailed description of need for change if no related issue -->

If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [x] I have opened a PR for the related docs change (if not, why?)
[docs PR](https://github.com/edgexfoundry/edgex-docs/pull/990)

## Testing Instructions
<!-- How can the reviewers test your change? -->
**Non-secure mode testing **
1. set the environment variable for non-secure testing `export EDGEX_SECURITY_SECRET_STORE=false`
2. run the latest compose builder `make run no-secty` to start core services
3. stop core data
4. remove the core data configuration in consul
5. modify local go-mod in edgex-go to use bootstrap from [this branch](https://github.com/ejlee3/go-mod-bootstrap/tree/common-config-flag)
7. build core data `make data`
9. run core-data locally and ensure that the configuration loads `cd cmd/core-data` and `./core-data -cc ../core-common-config-bootstrapper/res/configuration.yaml`
10. Verify that the core data configuration and common config were loaded using curl `curl http://localhost:59880/api/v2/config | json_pp`

**secure mode testing **
Repeat the above steps with `EDGEX_SECRUITY_SECRETS_STORE=true`, run all services in secure mode, and run core data with `sudo` 

## New Dependency Instructions (If applicable)
<!-- Please follow [vetting instructions](https://wiki.edgexfoundry.org/display/FA/Vetting+Process+for+3rd+Party+Dependencies) and place results here -->

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -180,6 +181,13 @@ func (cp *Processor) Process(
 
 	// Now must load configuration from local file if any of these conditions are true
 	if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {
```
```suggestion
	if !useProvider {
```

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -180,6 +181,13 @@ func (cp *Processor) Process(
 	if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {
+		commonConfigLocation := environment.GetCommonConfig(cp.lc, cp.flags.CommonConfig())
+		if commonConfigLocation != "" {
+			err := cp.loadCommonConfigFromUri(commonConfigLocation, serviceConfig, serviceType)
```
Lets not use UIR term yet. When we do implement URI if will be a generic function not specific to configuration. My thought is this loadCommonConfigFromFile will can it to do the actually loading once we have it in place.
```suggestion
			err := cp.loadCommonConfigFromFile(commonConfigLocation, serviceConfig, serviceType)
```

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -180,6 +181,13 @@ func (cp *Processor) Process(
+		if commonConfigLocation != "" {
+			err := cp.loadCommonConfigFromUri(commonConfigLocation, serviceConfig, serviceType)
+			if err != nil {
+				return fmt.Errorf("could not load oommon config: %s", err.Error())
```
error already has context
```suggestion
				return err
```

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
+
+	var err error
+
+	//TODO: implement reading from URI
```
The TODO can be removed

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
+	// separate out the necessary sections
+	allServicesConfig, ok := commonConfig["all-services"].(map[string]any)
+	if !ok {
+		return fmt.Errorf("could not find all-services section in config %s", configFile)
```
```suggestion
		return fmt.Errorf("could not find all-services section in common config %s", configFile)
```

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
+		cp.lc.Infof("loading the common configuration for service type %s", serviceType)
+		serviceTypeConfig, ok = commonConfig[appServicesKey].(map[string]any)
+		if !ok {
+			return fmt.Errorf("could not find %s section in config %s", appServicesKey, configFile)
```
```suggestion
			return fmt.Errorf("could not find %s section in common config %s", appServicesKey, configFile)
```

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
+		cp.lc.Infof("loading the common configuration for service type %s", serviceType)
+		serviceTypeConfig, ok = commonConfig[deviceServicesKey].(map[string]any)
+		if !ok {
+			return fmt.Errorf("could not find %s section in config %s", deviceServicesKey, configFile)
```
```suggestion
			return fmt.Errorf("could not find %s section in common config %s", deviceServicesKey, configFile)
```

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
+	//TODO: implement reading from URI
+	commonConfig, err := cp.loadConfigYamlFromFile(configFile)
+	if err != nil {
+		return fmt.Errorf("could not load common configuration %s: %s", configFile, err.Error())
```
error already has context
```suggestion
		return err
```

lenny-intel commented on bootstrap/config/config_test.go:
```diff
@@ -246,6 +247,54 @@ func TestLoadCommonConfig(t *testing.T) {
+			require.NotNil(t, cancel)
+			if tc.expectedErr == "" {
+				assert.NoError(t, err)
+				assert.NotEmpty(t, tc.serviceType)
```
Why is this assert needed? It is part of the tests code, not the code being tested.

lenny-intel commented on bootstrap/environment/variables.go:
```diff
@@ -358,15 +359,27 @@ func GetProfileDir(lc logger.LoggingClient, profileDir string) string {
 
+// GetCommonConfig gets the common configuration value from the Variables value (if it exists)
+// or uses passed in value.
+func GetCommonConfig(lc logger.LoggingClient, commonConfig string) string {
```
```suggestion
func GetCommonConfigFileName(lc logger.LoggingClient, commonConfigFileName string) string {
```

lenny-intel commented on bootstrap/environment/variables_test.go:
```diff
@@ -238,6 +238,35 @@ func TestGetConfigFileName(t *testing.T) {
 	}
 }
 
+func TestGetCommonConfig(t *testing.T) {
```
```suggestion
func TestGetCommonConfigFileName(t *testing.T) {
```

lenny-intel commented on bootstrap/flags/flags.go:
```diff
@@ -133,6 +137,11 @@ func (d *Default) ConfigFileName() string {
 	return d.configFileName
 }
 
+// CommonConfig returns the uri for the common configuration
```
```suggestion
// CommonConfig returns the location for the common configuration
```

lenny-intel commented on bootstrap/flags/flags.go:
```diff
@@ -145,6 +154,8 @@ func (d *Default) helpCallback() {
 			"Server Options:\n"+
 			"    -cp, --configProvider           Indicates to use Configuration Provider service at specified URL.\n"+
 			"                                    URL Format: {type}.{protocol}://{host}:{port} ex: consul.http://localhost:8500\n"+
+			"    -cc, --commonConfig             Takes the URI that specifies where the common configuration is pulled from when\n"+
```
```suggestion
			"    -cc, --commonConfig             Takes the location that specifies where the common configuration is loaded from when\n"+
```

lenny-intel commented on bootstrap/flags/flags.go:
```diff
@@ -154,7 +154,7 @@ func (d *Default) helpCallback() {
 			"    -cp, --configProvider           Indicates to use Configuration Provider service at specified URL.\n"+
 			"                                    URL Format: {type}.{protocol}://{host}:{port} ex: consul.http://localhost:8500\n"+
-			"    -cc, --commonConfig             Takes the URI that specifies where the common configuration is pulled from when\n"+
+			"    -cc, --commonConfig             Takes the location to specifies where the common configuration is loaded from when\n"+
```
Takes the location to where the common configuration is loaded ....

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -176,10 +177,15 @@ func (cp *Processor) Process(
-
-	// Now must load configuration from local file if any of these conditions are true
-	if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {
+	} else {
+		// Now must load configuration from local file if any of these conditions are true
+		commonConfigLocation := environment.GetCommonConfigFileName(cp.lc, cp.flags.CommonConfig())
+		if commonConfigLocation != "" {
+			err := cp.loadCommonConfigFromFile(commonConfigLocation, serviceConfig, serviceType)
+			if err != nil {
+				return err
+			}
+		}
```
```suggestion
	} else {
		// Now must load configuration from local file if any of these conditions are true
		commonConfigLocation := environment.GetCommonConfigFileName(cp.lc, cp.flags.CommonConfig())
		if commonConfigLocation != "" {
			err := cp.loadCommonConfigFromFile(commonConfigLocation, serviceConfig, serviceType)
			if err != nil {
				return err
			}
		}
	}
	
       	// Now must load private configuration from local file if any of these conditions are true
       if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {```

lenny-intel commented on bootstrap/config/config.go:
```diff
@@ -176,33 +177,42 @@ func (cp *Processor) Process(
-		}
-		cp.lc.Infof("Configuration loaded from file with %d overrides applied", overrideCount)
+		// Now load the private config from a local file if any of these conditions are true
+		if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {
```
Move outside the else cause

lenny-intel reviewed (approved):
LGTM.

MonicaisHer commented:
I updated the go-mod in edgex-go to use the bootstrap from this PR, but encountered some errors while building edgex-go. Here are the build steps and error details:

1. I added `replace github.com/edgexfoundry/go-mod-bootstrap/v3 v3.0.0-dev.54 => github.com/ejlee3/go-mod-bootstrap/v3 v3.0.0-20230320232408-c3ed78377b49` to the go.mod file and ran `go mod tidy`
2. I ran `snapcraft` to build the snapped binary
However, I received the following error:
```
internal/security/config/command/flags_common.go:37:9: cannot use &commonFlags (value of type *commonFlags) as flags.Common value in return statement: *commonFlags does not implement flags.Common (missing method CommonConfig)
make: *** [Makefile:134: cmd/secrets-config/secrets-config] Error 1
```
I'm wondering if there are any adjustments that need to be made to the edgex-go code as well?

cloudxxx8 commented on bootstrap/config/config.go:
```diff
@@ -310,6 +326,53 @@ func (cp *Processor) loadCommonConfig(
+		return err
+	}
+	// separate out the necessary sections
+	allServicesConfig, ok := commonConfig["all-services"].(map[string]any)
```
use `allServicesKey`

cloudxxx8 commented on bootstrap/config/config.go:
```diff
@@ -527,10 +590,27 @@ func (cp *Processor) loadPrivateFromFile() (*toml.Tree, error) {
 }
 
+// loadConfigYamlFromFile attempts to read the configuration yaml file
+func (cp *Processor) loadConfigYamlFromFile(yamlFile string) (map[string]interface{}, error) {
```
use `map[string]any` instead of `map[string]interface{}`

cloudxxx8 commented on bootstrap/config/config.go:
```diff
@@ -527,10 +590,27 @@ func (cp *Processor) loadPrivateFromFile() (*toml.Tree, error) {
+		return nil, fmt.Errorf("failed to read common configuration file %s: %s", yamlFile, err.Error())
+	}
+
+	data := make(map[string]interface{})
```
use `map[string]any` instead of `map[string]interface{}`

lenny-intel commented:
> I updated the go-mod in edgex-go to use the bootstrap from this PR, but encountered some errors while building edgex-go. Here are the build steps and error details:
> 
> 1. added `replace github.com/edgexfoundry/go-mod-bootstrap/v3 v3.0.0-dev.54 => github.com/ejlee3/go-mod-bootstrap/v3 v3.0.0-20230320232408-c3ed78377b49` to the go.mod file and ran `go mod tidy`
> 2. ran `snapcraft` to build the snapped binary
> 
> However, I received the following error:
> 
> ```
> internal/security/config/command/flags_common.go:37:9: cannot use &commonFlags (value of type *commonFlags) as flags.Common value in return statement: *commonFlags does not implement flags.Common (missing method CommonConfig)
> make: *** [Makefile:134: cmd/secrets-config/secrets-config] Error 1
> ```
> 
> I'm wondering if there are any adjustments that need to be made to the edgex-go code as well?

@MonicaisHer Yes, edgex-go will need tweaks for those services that re-implement the flags.Common interface. They need to implement the CommonConfig API.

lenny-intel reviewed (approved):
LGTM

ejlee3 commented:
> > I updated the go-mod in edgex-go to use the bootstrap from this PR, but encountered some errors while building edgex-go. Here are the build steps and error details:
> > 
> > 1. added `replace github.com/edgexfoundry/go-mod-bootstrap/v3 v3.0.0-dev.54 => github.com/ejlee3/go-mod-bootstrap/v3 v3.0.0-20230320232408-c3ed78377b49` to the go.mod file and ran `go mod tidy`
> > 2. ran `snapcraft` to build the snapped binary
> > 
> > However, I received the following error:
> > ```
> > internal/security/config/command/flags_common.go:37:9: cannot use &commonFlags (value of type *commonFlags) as flags.Common value in return statement: *commonFlags does not implement flags.Common (missing method CommonConfig)
> > make: *** [Makefile:134: cmd/secrets-config/secrets-config] Error 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm wondering if there are any adjustments that need to be made to the edgex-go code as well?
> 
> @MonicaisHer Yes, edgex-go will need tweaks for those services that re-implement the flags.Common interface. They need to implement the CommonConfig API.

Draft PR for edgex-go changes is here: https://github.com/edgexfoundry/edgex-go/pull/4464

lenny-intel commented:
@MonicaisHer , please let us know once you have verified in edgex-go snap so we can merge this PR. THX!

cloudxxx8 reviewed (approved):
LGTM

MonicaisHer reviewed (approved):
Thank you! I tested this [PR](https://github.com/edgexfoundry/edgex-go/pull/4464) with snap to use a common config option, and it worked as expected. I was able to choose not to use the config provider and instead pull the common config from a file. The environment overrides were then applied after that. Here are the steps I used:
```
# Install snap built from this PR's changes
$ snap install --dangerous ./edgexfoundry_3.0.0-dev.108_amd64.snap

# Set not to use config provider
sudo snap set edgexfoundry apps.core-data.config.edgex-config-provider='none'

# Pull common config
sudo snap set edgexfoundry apps.core-data.config.edgex-common-config=./config/core-common-config-bootstrapper/res/configuration.yaml

# Set test environment variables
sudo snap set edgexfoundry apps.core-data.config.service-startupmsg='manual-snap-testing'

snap restart edgexfoundry.core-data
```
To verify that the common config was pulled and the environment variables were applied:
```
$ journalctl -f | grep core-data
app=core-data source=variables.go:402 msg="Variables override of 'Configuration Provider Information' by environment variable: EDGEX_CONFIG_PROVIDER=none"
app=core-data source=variables.go:402 msg="Variables override of '-cc/--commonConfig' by environment variable: EDGEX_COMMON_CONFIG=./config/core-common-config-bootstrapper/res/configuration.yaml
app=core-data source=config.go:599 msg="reading ./config/core-common-config-bootstrapper/res/configuration.yaml"
app=core-data source=config.go:194 msg="Common configuration loaded from file with 1 overrides applied"
app=core-data source=message.go:51 msg="Starting core-data 3.0.0-dev.108 "
app=core-data source=message.go:55 msg="manual-snap-testing"
```
